### ISSUE TITLE
deno 1.11.0 TypeScript compilation issue

### ISSUE THREAD

**Issue by jdfoster** [denoland/deno#10916]
I am having some issues compiling TypeScript in v1.11.0, that I have not observed with v1.10.3. With the latest release I am having problems compiling the simplest Zod example (see below), but I have also had similar issues with the yaml decoder module from the standard library. For completeness, my system is running Ubuntu 20.04.2 LTS (kernel 5.8.0-55-generic) and am happy to provide any additional information on request.

```
import { z } from "https://deno.land/x/zod@v3.1.0/mod.ts";

const mySchema = z.string();
const value = mySchema.parse("tuna");
console.log(value);

```

In v1.10.3, running `deno cache --log-level debug example.ts` works beautifully.

```
...OMITTED FOR BREVITY...
DEBUG RS - deno::program_state:264 - Compilation statistics:
  Files: 65
  Nodes: 51743
  Identifiers: 19057
  Symbols: 31777
  Types: 17345
  Instantiations: 38620
  Parse time: 139
  Bind time: 106
  Check time: 1083
  Emit time: 0
  Total TS time: 1328
  Compile time: 1342
```

Whereas on v1.11.0, running `deno cache --log-level debug example.ts` throw the following errors.

```
...OMITTED FOR BREVITY...
DEBUG RS - deno::program_state:268 - Compilation statistics:
  Files: 67
  Nodes: 52206
  Identifiers: 19221
  Symbols: 3230760
  Types: 2196061
  Instantiations: 6247049
  Parse time: 132
  Bind time: 97
  Check time: 40124
  Emit time: 187
  Total TS time: 40540
  Compile time: 40553

error: TS2321 [ERROR]: Excessive stack depth comparing types 'ZodError<?>' and 'ZodError<?>'.
    processError(this);
    ~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/ZodError.ts:179:5

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodFormattedError<?>' and 'ZodFormattedError<?>'.
    processError(this);
    ~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/ZodError.ts:179:5

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodIntersection<?, U>' and 'ZodIntersection<?, U>'.
      ? ZodObject<
        ^
    at https://deno.land/x/zod@v3.1.0/helpers/partialUtil.ts:10:9

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodIntersection<T, ?>' and 'ZodIntersection<T, ?>'.
      ? ZodObject<
        ^
    at https://deno.land/x/zod@v3.1.0/helpers/partialUtil.ts:10:9

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodNonEmptyArray<?>' and 'ZodNonEmptyArray<?>'.
      ? ZodObject<
        ^
    at https://deno.land/x/zod@v3.1.0/helpers/partialUtil.ts:10:9

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodOptional<?>' and 'ZodOptional<?>'.
      ? ZodObject<
        ^
    at https://deno.land/x/zod@v3.1.0/helpers/partialUtil.ts:10:9

TS2577 [ERROR]: Return type annotation circularly references itself.
  ) => ZodEffectsType<This> = (check, message) => {
       ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/types.ts:204:8

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodEffects<?, Output>' and 'ZodEffects<?, Output>'.
      return this._refinement((val, ctx) => {
             ^
    at https://deno.land/x/zod@v3.1.0/types.ts:206:14

TS2577 [ERROR]: Return type annotation circularly references itself.
  ) => ZodEffectsType<This> = (check, refinementData) => {
       ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/types.ts:268:8

TS2577 [ERROR]: Return type annotation circularly references itself.
  ): ZodEffectsType<This> {
     ~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/types.ts:293:6

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodNullable<?>' and 'ZodNullable<?>'.
  > = () => this.optional().nullable();
            ~~~~~~~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/types.ts:325:13

TS2577 [ERROR]: Return type annotation circularly references itself.
  ): This extends ZodUnion<infer Opts>
     ^
    at https://deno.land/x/zod@v3.1.0/types.ts:331:6

TS2577 [ERROR]: Return type annotation circularly references itself.
  ): This extends ZodEffects<infer T, any>
     ^
    at https://deno.land/x/zod@v3.1.0/types.ts:348:6

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodUnion<?>' and 'ZodUnion<?>'.
    return new ZodUnion({
           ^
    at https://deno.land/x/zod@v3.1.0/types.ts:1817:12

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodTuple<?>' and 'ZodTuple<?>'.
export interface ZodFunctionDef<
                 ~~~~~~~~~~~~~~
    at https://deno.land/x/zod@v3.1.0/types.ts:2179:18

TS2321 [ERROR]: Excessive stack depth comparing types 'ZodLazy<?>' and 'ZodLazy<?>'.
    return new ZodLazy({
           ^
    at https://deno.land/x/zod@v3.1.0/types.ts:2364:12

Found 16 errors.
```

Thanks for sterling work, I absolutely adore deno.

**Comment by dsherret** [comment]
See https://github.com/microsoft/TypeScript/issues/44299#issuecomment-853063196 -- other people having trouble with Zod.

**Comment by jdfoster** [comment]
@dsherret Thanks for the links. The incompatibility between zod and typescript has been patched in the latest zod release (v3.2.0). I shall close this issue. Thanks for the great work.
